File: native/app/group_commands.go
package app

import (
	"errors"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupauthority"
	"github.com/JohnnyPBelo/relayloom/native/groupledger"
	"github.com/JohnnyPBelo/relayloom/native/groups"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"github.com/JohnnyPBelo/relayloom/native/profilestate"
)

// Called with the application mutex held. No result or network side effect
// escapes before the outer profile transaction commits. This is management;
// dynamic message admission and control-carrier synchronization follow separately.
func (n *Node) groupCommandLocked(body map[string]any) (any, error) {
	if n.identity == nil || n.privateDatabase == nil {
		return nil, errors.New("desbloqueie a identidade")
	}
	if !contains([]string{"list", "state", "operation", "proofs", "private-state", "headers", "snapshot", "remember", "leave"}, text(body["action"])) {
		if err := n.requireGroupReplayLocked(); err != nil {
			return nil, err
		}
	}
	update := n.updateGroupState
	if update == nil {
		update = n.privateDatabase.Update
	}
	var result any
	next, err := copyPrivate(n.private, n.identity.Public.ID)
	if err != nil {
		return nil, err
	}
	var decisions map[string]groupRetry
	err = update(func(tx *groupstore.Tx) error {
		current, err := profilestate.Read(tx)
		if err != nil {
			return err
		}
		if current == nil || current.Digest != n.privateDigest {
			return errors.New("o estado privado mudou; volte a lê-lo antes de gerir o grupo")
		}
		_, err = groupledger.Run(tx, *n.identity, func(l *groupledger.Ledger, g *groupauthority.Registry) error {
			var err error
			result, err = executeGroupCommand(g, n.identity.Public, body)
			if err != nil {
				return err
			}
			decisions, err = reconcileGroupOutbox(l, next.Outbox)
			return err
		})
		return err
	})
	if err != nil {
		// A failed response can follow a successful SQL commit. Recover through
		// the signed binding, never by restoring an earlier private JSON file.
		identity := *n.identity
		_ = n.privateDatabase.Close()
		database, recovered, digest, readErr := openPrivateProfile(n.Dir, identity)
		if readErr == nil {
			n.privateDatabase, n.private, n.privateDigest = database, recovered, digest
			if err := n.restoreGroupHoldsLocked(); err != nil {
				_ = n.lockPrivateLocked()
			}
		} else {
			_ = n.lockPrivateLocked()
		}
		return nil, err
	}
	n.private, n.groupRetries = next, decisions
	if err := n.restoreGroupHoldsLocked(); err != nil {
		n.recoverGroupContentLocked()
		return nil, err
	}
	n.groupSync.record(body, result)
	return result, nil
}

func executeGroupCommand(g *groupauthority.Registry, identity core.PublicIdentity, body map[string]any) (any, error) {
	action, err := fieldString(body, "action")
	if err != nil {
		return nil, err
	}
	id, operation, expected := text(body["groupId"]), text(body["operationId"]), text(body["expected"])
	var result groupauthority.Result
	switch action {
	case "list":
		list, err := g.List()
		return map[string]any{"groups": list, "limits": map[string]int{"groups": groupauthority.GroupLimit, "operations": groupauthority.OperationLimit}, "management": true, "messaging": false}, err
	case "state":
		state, err := g.State(id)
		return map[string]any{"group": state}, err
	case "operation":
		status, err := g.OperationStatus(operation)
		return map[string]any{"operation": status}, err
	case "create":
		title, err := fieldString(body, "title")
		if err != nil {
			return nil, err
		}
		result, err = g.Create(operation, title)
		if err != nil {
			return nil, err
		}
	case "invite":
		card, err := publicIdentity(body["card"])
		if err != nil {
			return nil, err
		}
		result, err = g.Invite(operation, id, expected, card)
		if err != nil {
			return nil, err
		}
	case "remember":
		raw, err := core.Canonical(body["anchor"])
		if err != nil {
			return nil, err
		}
		anchor, err := groups.DecodeAnchor(raw)
		if err != nil {
			return nil, err
		}
		raw, err = core.Canonical(body["parent"])
		if err != nil {
			return nil, err
		}
		parent, err := groups.DecodeEpoch(raw, anchor)
		if err != nil {
			return nil, err
		}
		raw, err = core.Canonical(body["invitation"])
		if err != nil {
			return nil, err
		}
		invitation, err := groups.DecodeInvitation(raw, anchor, parent, identity)
		if err != nil {
			return nil, err
		}
		result, err = g.RememberInvitation(operation, anchor, parent, invitation)
		if err != nil {
			return nil, err
		}
	case "accept":
		result, err = g.Accept(operation, id, expected)
	case "commit":
		var update groups.Update
		update, err = parseGroupUpdate(body)
		if err == nil {
			result, err = g.Commit(operation, id, expected, update)
		}
	case "close":
		result, err = g.Close(operation, id, expected)
	case "leave":
		result, err = g.Leave(operation, id)
	case "headers":
		raw, err := core.Canonical(body["headers"])
		if err != nil {
			return nil, err
		}
		observed, err := g.ObserveRawHeaders(id, raw)
		return map[string]any{"observation": observed}, err
	case "snapshot":
		raw, err := core.Canonical(body["snapshot"])
		if err != nil {
			return nil, err
		}
		snapshot, err := groups.ParseSnapshot(raw)
		if err != nil {
			return nil, err
		}
		state, err := g.ObserveSnapshot(id, text(body["epochId"]), snapshot)
		return map[string]any{"group": state}, err
	case "resume":
		state, err := g.ResumeCapacity(id)
		return map[string]any{"group": state}, err
	case "proofs":
		from, err := number(body["from"])
		if err != nil || from < 0 || from >= groups.EpochLimit {
			return nil, errors.New("pedido de provas fora dos limites")
		}
		count := int64(groupauthority.PageLimit)
		if value, exists := body["count"]; exists {
			count, err = number(value)
			if err != nil || count < 1 || count > groupauthority.PageLimit {
				return nil, errors.New("pedido de provas fora dos limites")
			}
		}
		anchor, err := g.Anchor(id)
		if err != nil {
			return nil, err
		}
		headers, err := g.Proofs(id, int(from), int(count))
		return map[string]any{"anchor": anchor, "headers": headers}, err
	case "private-state":
		snapshot, err := g.PrivateState(id, text(body["epochId"]))
		return map[string]any{"snapshot": snapshot}, err
	default:
		return nil, errors.New("acção de grupo desconhecida")
	}
	if err != nil {
		return nil, err
	}
	state, err := g.State(result.GroupID)
	return map[string]any{"operation": result, "group": state}, err
}

func parseGroupUpdate(body map[string]any) (groups.Update, error) {
	var result groups.Update
	title, err := fieldString(body, "title")
	if err != nil {
		return result, err
	}
	roster, err := members(body["members"])
	if err != nil {
		return result, err
	}
	items, ok := body["joins"].([]any)
	if !ok || len(items) > groups.MemberLimit {
		return result, errors.New("consentimentos fora dos limites")
	}
	joins := make([]groups.GroupConsent, 0, len(items))
	for _, item := range items {
		raw, err := core.Canonical(item)
		if err != nil {
			return result, err
		}
		consent, err := groups.ParseConsent(raw)
		if err != nil {
			return result, err
		}
		joins = append(joins, consent)
	}
	return groups.Update{Title: title, Members: roster, Joins: joins}, nil
}

File: apps/node/src/group-commands.ts
import type { Identity } from "../../../packages/core/src/index.js";
import type { ProfileDatabase } from "../../../packages/profile/src/database.js";
import { readProfileState } from "../../../packages/profile/src/state.js";
import { GroupLedger } from "../../../packages/groups/src/ledger.js";
import type { GroupOperationResult } from "../../../packages/groups/src/registry.js";
import type {
  GroupAnchor,
  GroupConsent,
  GroupEpoch,
  GroupInvitation,
  GroupSnapshot,
} from "../../../packages/groups/src/certificates.js";
import type { PublicIdentity } from "../../../packages/core/src/index.js";
import {
  GroupNotices,
  type GroupNotice,
} from "../../../packages/groups/src/notices.js";

export type GroupCommand =
  | { action: "list" }
  | { action: "notice-list" | "notice-outbox" }
  | { action: "notice-dismiss"; id: string }
  | { action: "notice-open"; id: string; operationId: string }
  | { action: "state"; groupId: string }
  | { action: "operation"; operationId: string }
  | { action: "create"; operationId: string; title: string }
  | {
      action: "invite";
      operationId: string;
      groupId: string;
      expected: string;
      card: PublicIdentity;
    }
  | {
      action: "remember";
      operationId: string;
      anchor: GroupAnchor;
      parent: GroupEpoch;
      invitation: GroupInvitation;
    }
  | {
      action: "accept" | "close";
      operationId: string;
      groupId: string;
      expected: string;
    }
  | { action: "leave"; operationId: string; groupId: string }
  | {
      action: "commit";
      operationId: string;
      groupId: string;
      expected: string;
      title: string;
      members: PublicIdentity[];
      joins: GroupConsent[];
    }
  | { action: "headers"; groupId: string; headers: GroupEpoch[] }
  | {
      action: "snapshot";
      groupId: string;
      epochId: string;
      snapshot: GroupSnapshot;
    }
  | { action: "resume"; groupId: string }
  | { action: "proofs"; groupId: string; from: number; count?: number }
  | { action: "private-state"; groupId: string; epochId: string };

/** The authenticated application API owns this command boundary. Signing
 * checks/CAS happen inside the existing profile transaction. No network bytes
 * or successful response leave this function before the outer commit. */
export function executeGroupCommand(
  database: ProfileDatabase,
  identity: Identity,
  privateDigest: string,
  command: GroupCommand,
  reconcile?: (ledger: GroupLedger) => void,
) {
  if (!command || typeof command !== "object" || Array.isArray(command))
    throw new Error("Comando de grupo inválido");
  return database.transaction((tx) => {
    if (readProfileState(tx)?.digest !== privateDigest)
      throw new Error(
        "O estado privado mudou; volte a lê-lo antes de gerir o grupo",
      );
    return GroupLedger.run(tx, identity, (ledger, g) => {
      const notices = new GroupNotices(tx, identity.public);
      const response = (() => {
        let result: unknown;
        switch (command.action) {
          case "notice-list":
            return { notices: notices.list("in") };
          case "notice-outbox":
            return { notices: notices.list("out") };
          case "notice-dismiss":
            return { retired: notices.retire("in", command.id) };
          case "notice-open": {
            const previous = g.operationStatus(command.operationId);
            if (
              previous?.certificate?.id === command.id &&
              previous.certificate.body.domain ===
                "relayloom/group-invitation/1" &&
              previous.certificate.body.inviteeId === identity.public.id
            ) {
              result = previous;
              break;
            }
            const notice = notices
              .list("in")
              .find((e) => e.notice.certificate.id === command.id)?.notice;
            if (!notice || notice.kind !== "invitation")
              throw new Error("Convite indisponível");
            const known = g.list().find((v) => v.id === notice.anchor.id);
            if (known) {
              const info = g.syncState(known.id);
              if (
                info.admitted ||
                (info.consent && info.invitation?.id === notice.certificate.id)
              )
                throw new Error(
                  "Convite já aceite; aguarda a confirmação ou consulta o grupo",
                );
            }
            if (
              known?.head &&
              known.head.body.number >= notice.parent.body.number &&
              known.head.id !== notice.parent.id
            )
              throw new Error(
                "O grupo já avançou; é necessário um convite actual",
              );
            result = g.rememberInvitation(
              command.operationId,
              notice.anchor,
              notice.parent,
              notice.certificate,
            );
            // Explicit opening makes this a known group. Its signed birth
            // header is sufficient by itself; later parents still need ancestry.
            g.observeHeaders(notice.anchor.id, [notice.parent]);
            break;
          }
          case "list":
            return {
              groups: g.list(),
              limits: { groups: 64, operations: 256 },
              management: true,
              messaging: false,
              noticeCount: notices.count("in"),
            };
          case "state":
            return { group: g.state(command.groupId) };
          case "operation":
            return { operation: g.operationStatus(command.operationId) };
          case "create":
            result = g.create(command.operationId, command.title);
            break;
          case "invite":
            result = g.invite(
              command.operationId,
              command.groupId,
              command.expected,
              command.card,
            );
            // Keep the exact delivery card with the issuance transaction.
            // Replaying an old operation must not revive an obsolete invitation.
            const parent = g.state(command.groupId).head;
            if (parent?.id === command.expected) {
              const notice: GroupNotice = {
                type: "group-notice",
                version: 1,
                kind: "invitation",
                anchor: g.anchor(command.groupId),
                parent,
                member: command.card,
                certificate: (result as GroupOperationResult)
                  .certificate as GroupInvitation,
              };
              notices.save("out", notice);
            }
            break;
          case "remember":
            result = g.rememberInvitation(
              command.operationId,
              command.anchor,
              command.parent,
              command.invitation,
            );
            break;
          case "accept":
            result = g.accept(
              command.operationId,
              command.groupId,
              command.expected,
            );
            break;
          case "commit":
            result = g.commit(
              command.operationId,
              command.groupId,
              command.expected,
              {
                title: command.title,
                members: command.members,
                joins: command.joins,
              },
            );
            break;
          case "close":
            result = g.close(
              command.operationId,
              command.groupId,
              command.expected,
            );
            break;
          case "leave":
            result = g.leave(command.operationId, command.groupId);
            break;
          case "headers":
            return {
              observation: g.observeHeaders(command.groupId, command.headers),
            };
          case "snapshot":
            return {
              group: g.observeSnapshot(
                command.groupId,
                command.epochId,
                command.snapshot,
              ),
            };
          case "resume":
            return { group: g.resumeCapacity(command.groupId) };
          case "proofs":
            return {
              anchor: g.anchor(command.groupId),
              headers: g.proofs(command.groupId, command.from, command.count),
            };
          case "private-state":
            return {
              snapshot: g.privateState(command.groupId, command.epochId),
            };
          default:
            throw new Error("Acção de grupo desconhecida");
        }
        return {
          operation: result,
          group: g.state((result as GroupOperationResult).groupId),
        };
      })();
      reconcile?.(ledger);
      return response;
    }).value;
  });
}
